File: 3/golang/main.go
// The prime factors of 13195 are 5, 7, 13 and 29.
//
// What is the largest prime factor of the number 600851475143 ?

package main

import (
    "fmt"
    "math"
)

func isPrime(n int) bool {
    for x := 2; x < int(math.Sqrt(float64(n))+1); x++ {
        if n % x == 0 {
            return false
        }
    }
    return true
}

func getPrimeFactorization(n int) []int {
    primeFactors := make([]int, 1)
    quotient := n
    current := 2
    for !isPrime(quotient) {
        if isPrime(current) && quotient % current == 0 {
            primeFactors = append(primeFactors, current)
            quotient = quotient / current
        }
        current += 1
    }
    return append(primeFactors, quotient)
}

func main() {
    number := 600851475143
    primeFactors := getPrimeFactorization(number)
    fmt.Println(primeFactors[len(primeFactors)-1])
}

File: 3/python/main.py
# The prime factors of 13195 are 5, 7, 13 and 29.
#
# What is the largest prime factor of the number 600851475143 ?

import math

number = 600851475143

def is_prime(n):
    for x in range(2, int(math.sqrt(n))+1):
        if n % x == 0:
            return False
    return True

def get_prime_factorization(n):
    prime_factors = []
    quotient = n
    current = 2
    while not is_prime(quotient):
        if is_prime(current) and quotient % current == 0:
            prime_factors.append(current)
            quotient = quotient // current
        current += 1
    prime_factors.append(quotient)
    return prime_factors

print(get_prime_factorization(number)[-1:][0])
